tests/Client/UpdaterTest.php

```diff
@@ -382,4 +389,57 @@ public function providerFileNotFoundExceptions():array
             ],
         ]);
     }
+
+    /**
+     * Tests that exceptions are thrown when a repo is in a rollback attack state.
+     *
+     * @param string $fixturesSet
+     *   The fixtures set.
+     * @param \Exception $expectedException
+     *   The expected exception.
+     * @param array $expectedUpdatedVersions
+     *   The expected repo file version after refresh attempt.
+     *
+     * @return void
+     *
+     * @dataProvider providerAttackRepoException
+     */
+    public function testAttackRepoException(string $fixturesSet, \Exception $expectedException, array $expectedUpdatedVersions): void
+    {
+        // Use the memory storage used so tests can write without permanent
+        // side-effects.
+        $this->localRepo = $this->memoryStorageFromFixture($fixturesSet, 'tufclient/tufrepo/metadata/current');
+        $this->testRepo = new TestRepo($fixturesSet);
+        $this->assertClientRepoVersions(static::getFixtureClientStartVersions($fixturesSet));
+        $updater = $this->getSystemInTest();
+        try {
+            $updater->refresh();
```

Comment:
If I understand this correctly, the rollback attack should be detected, and an exception should be immediately thrown, before any further downloads take place. Right? If that's the case, would it maybe be worth it to find a way to test that the attack is detected *when* we expect it to be, as opposed to "well, eventually a rollback attack is detected, and an exception is thrown", which is what it looks like we are currently doing.

I say this because I know the updater continually is writing things to persistent storage as it refreshes itself. So in theory, if a bad guy could poison the persistent storage _before_ a rollback attack is detected, that could potentially lay the groundwork for further (successful) attacks.

Reply:
Nice idea. IMHO we should move `setExceptionOnChange()` into whatever the interface is for the durable storage backends, and rename it to `lock()` (and add an `unlock()` method too). Can be done in a later PR, though.

Reply:
I think if we only need this for testing we should not add it an interface. We should be able to refactor our test with changing our non-test code and affecting any API we may be creating. 

I have added a test for `setExceptionOnChange()`